### ISSUE TITLE
Won't compile with LLVM 3.6

### ISSUE THREAD

**Issue by olleharstedt** [whitequark/ocaml-llvm-ng#1]
Hi!

I'm trying to install this with LLVM 3.6, but it can't find package llvm_X86:

~/kod/ocaml/ocaml-llvm-ng$ make all
ocamlbuild -j 4 -use-ocamlfind lib_test/run_cmm.native
Finished, 0 targets (0 cached) in 00:00:00.
 ocamlfind ocamldep -package compiler-libs.optcomp -package llvm -package llvm.target -package llvm_X86 -modules lib_test/run_cmm.ml > lib_test/run_cmm.ml.depends
ocamlfind: Package `llvm_X86' not found
Command exited with code 2.
Compilation unsuccessful after building 1 target (0 cached) in 00:00:00.
make: **\* [all] Error 10
~/kod/ocaml/ocaml-llvm-ng$ 

This file seems to be used only up to 3.5?

http://www.rpmfind.net/linux/rpm2html/search.php?query=ocaml%28Llvm_X86%29

But installing LLVM via opam will get you version 3.6. So what's the recommended solution?

Regards
Olle

**Comment by whitequark** [comment]
If the llvm_X86 package is not present, the packagers have screwed up. Upstream builds it always from 3.5 onwards, assuming the corresponding backend is available.

**Comment by olleharstedt** [comment]
Seems like Llvm_X86 is a module within package Llvm.

Did you ever get the garbage-collector to work? This project doesn't seem to use gcread/gcwrite.

**Comment by whitequark** [comment]
Nope

**Comment by olleharstedt** [comment]
Sorry to bother you, but do you know of anyone or any project that got LLVM to work with the OCaml GC?

**Comment by whitequark** [comment]
No, and I also don't think it (combining LLVM and existing OCaml's GC) is a worthwhile goal. LLVM's old GC architecture (which can generate OCaml metadata) is excessively conservative: it only allows one to point out GC roots on stack and not in the registers. So while it would not be a large amount of work to integrate the two, the result will execute considerably slower than stock OCaml; gcroot intrinsics together with loads and stores to them act as blanket barriers that inhibit further optimizations.

You might have some success with LLVM's [new GC architecture](http://llvm.org/docs/StackMaps.html), contributed by Azul, but it will not work with OCaml's GC out of box and cannot be modified so that it would.

In addition, you will have to add several custom LLVM passes to teach LLVM the OCaml semantics, particularly so that it will be able to lift allocations. LLVM does not currently recognize `caml_alloc*` as allocation functions and will not turn them into allocas, which is the prerequisite for all interesting optimizations.

In addition, you will have to modify the Cmm generator so that it will emit fewer redundant casts, e.g. between ints and floats and especially ints and pointers. These casts confuse LLVM's optimization passes and inhibit instruction combining as well as lifting loads/stores.

At this point you should be roughly on par with ocamlopt. You might also reap some benefit from vectorization passes, but it is equally likely that you will have to teach LLVM the invariants that hold in OCaml-emitted loops, which are slightly different than in C-emitted loops.

In short, the challenge here is getting the GC to cooperate with the kind of LLVM-generated metadata that does not inhibit many optimizations, plus teaching LLVM the kind of assumptions about OCaml that it takes for granted when compiling C code.

**Comment by olleharstedt** [comment]
do you think it would be easier to integrate the ocaml gc if one disable some of its features, like compacting/copying?

i'm working on a statically typed subset of php not unlike your ruby compiler. do you want to tell me why you abannoded the project (no idea how your schedule look like)?

**Comment by whitequark** [comment]
You can't "disable features" of a GC, that's not how they work.

I got bored (with ocaml-llvm-ng). For the Ruby compiler, see http://whitequark.org/blog/2013/12/21/foundry-has-been-cancelled/

**Comment by olleharstedt** [comment]
Again, sorry for disturbing you. Ignore if you don't have time.

Did you ever consider generating C-code instead of LLVM IR, something like RPython? Then it would be easy to insert the needed OCaml C-macros for its GC, although you would miss many of the LLVM optimisation passes, I guess.

**Comment by whitequark** [comment]
It's not about macros. In fact, the invariants that the OCaml GC relies upon cannot be provided by C code, which is why gc.h provides the awful caml_local_roots hack. You need precise understanding of placement of values on stack and in registers, which C (obviously) doesn't give you, old LLVM GC gave you on stack with the right format, and new LLVM GC gives you entirely but in the wrong format.

Anything like caml_local_roots would inhibit optimizations. It's just pointless to write a backend that is inherently substantially slower than ocamlopt.

**Comment by olleharstedt** [comment]
I'm not competing with ocamlopt, but with PHP and HHVM. So maybe a hack could be used for now. I need much deeper understanding of garbage collection before I can do anything interesting with LLVM myself.

Anyway, thanks for your time!
